gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergCatalog.java

```diff
@@ -17,10 +17,25 @@
 
 package org.apache.gobblin.data.management.copy.iceberg;
 
+import org.apache.iceberg.catalog.Catalog;
+
 
 /**
  * Any catalog from which to access {@link IcebergTable}s.
  */
 public interface IcebergCatalog {
   IcebergTable openTable(String dbName, String tableName);
+  String getCatalogUri();
+
+  /**
+   * Adding a sub interface to help us provide an association between {@link Catalog} and {@link IcebergCatalog}.
+   * This helps us resolve to the Catalog to its concrete implementation class
+   * Primarily needed to access `newTableOps` method which only certain {@link Catalog} derived classes open for public access
+   */
+  interface CatalogSpecifier {
```

Comment:
why we need this specifier? Why not just two configs: one controls what real catalog we are using, and another control the gobblin catalog we are using?

Reply:
Yeah I'm also wondering on this, is this layer of abstraction necessary for future extensions?

Reply:
@ZihanLi58 @Will-Lo The interface is introduced primarily to bridge the gap between `IcebergCatalog` (exists in the Gobblin world as part of distcp) and the concrete implementation of the type of catalog for eg. `HiveCatalog` which implements `org.apache.iceberg.catalog`. Now, `newTableOps` is required for us to interact with the iceberg tables in order to perform distcp. But since the method `newTableOps` is protected, we need to resolve `Catalog` to its concrete implementation and also pair `Catalog` (exists in Iceberg world) with `IcebergCatalog` (exists in Gobblin world) to perform iceberg based distcp

Now, we can do this through configs as well, but essentially one config is needed i.e. the specifier and we can resolve everything else from the specific type of catalog. This reduces the hassle of adding additional props in the configs/job templates and also gives the liberty to extend the functionality in future when we support other catalog types and few functionality might be specific to that catalog type. Let me know if this helps!

Reply:
For other types of iceberg operation, it seems like gobblin icebergCatalog is not needed. And it does not make sense for me to declare a specifier if I just want to use a different type of catalog.  
Can we have only one Gobblin IcebergCatalog class that takes the normal catalog, and throw an exception if the normal catalog does not set newTableOps as a public method? You can use this https://www.digitalocean.com/community/tutorials/java-reflection-example-tutoria to check whether a method is protect or public

Reply:
If we are saying icebergCatalog's behavior is binding with the underlying catalog type, given that we already have one easy way (CatalogUtil.loadCatalog()) to load the underlying catalog easily, I think you can just keep with the original implementation, where you just specify the icebergCatalog type and then under icebergCatalog, you force the specific catalog (hiveCatalog for example). On IcebergMetataWriter side, we can just use CatalogUtil to load the catalog...

Reply:
I agree with using `CatalogUtils.loadCatalog()` method while creating the Catalogs based on their impl that's provided as arg and we are also using that in the `IcebergCatalogFactory`. We make use of `CatalogUtils` itself for loading catalogs based on their impl provided as part of the catalog specifier which taken as a config. 

We have `IcebergCatalog` interface to provide us with methods that will enable us to work with `IcebergTable` and translate them to gobblin constructs for distcp. Regarding, the `IcebergMetadataWriter` yes, we will use `CatalogUtils.loadCatalog()` provided by the Iceberg open-source project. 

@ZihanLi58 can you please elaborate on what you do mean by original implementation about specifying the IcebergCatalog. Because, originally, it was hard-coded to a `HiveCatalog` and to avoid that and provide a pairing between the catalogs in gobblin and iceberg world we have this specifier!

Reply:
Yes, I mean if we are sure that a specific GobblinIcebergCatalog only supports one type of native iceberg catalog as you mentioned above, i.e. IcebergHiveCatalog only supports HiveCatalog, then I think it's better to have IcebergHiveCatalog to hardcode HiveCatalog, and a job just needs to tell that it needs to use IcebergHiveCatalog. 
If we are saying one GobblinIcebergCatalog can support different types of native iceberg catalogs, then we should separate the config into two to make it more flexible. Let's say we have icebergCatalog, and if I want to use FsCatalog in the future, it does not make sense for me to create another specifier class that does nothing but just combines icebergCatalog and FsCatalog.

Reply:
Gotcha! I have updated the logic and refactored the classes after getting rid of `CatalogSpecifier`